pylate/losses/cached_contrastive.py

```diff
@@ -0,0 +1,348 @@
+        a RandContext so that we can exactly reproduce this forward pass later.
+        """
+        grad_context = nullcontext if with_grad else torch.no_grad
+        random_state_context = nullcontext() if random_state is None else random_state
+        sentence_feature_minibatch = {
+            k: v[begin:end] for k, v in sentence_feature.items()
+        }
+        with random_state_context:
+            with grad_context():
+                # If we need a new random-state copy, create it
+                random_state = (
+                    RandContext(*sentence_feature_minibatch.values())
+                    if copy_random_state
+                    else None
+                )
+                outputs = self.model(sentence_feature_minibatch)
+                # by default, PyLate ColBERT forward returns a dict with "token_embeddings"
+                embeddings = F.normalize(outputs["token_embeddings"], p=2, dim=-1)
+
+        return embeddings, random_state
+
+    def embed_minibatch_iter(
+        self,
+        sentence_feature: dict[str, Tensor],
+        with_grad: bool,
+        copy_random_state: bool,
+        random_states: list[RandContext] | None = None,
+    ) -> Iterator[tuple[Tensor, RandContext | None]]:
+        """
+        Yields chunks of embeddings (and corresponding RandContext) for the given
+        sentence_feature, respecting the mini_batch_size limit.
+        """
+        input_ids = sentence_feature["input_ids"]
+        bsz = input_ids.size(0)
+        for i, b in enumerate(
+            tqdm.trange(
+                0,
+                bsz,
+                self.mini_batch_size,
+                desc="Embed mini-batches",
+                disable=not self.show_progress_bar,
+            )
+        ):
+            e = b + self.mini_batch_size
+            reps, random_state = self.embed_minibatch(
+                sentence_feature=sentence_feature,
+                begin=b,
+                end=e,
+                with_grad=with_grad,
+                copy_random_state=copy_random_state,
+                random_state=None if random_states is None else random_states[i],
+            )
+            yield reps, random_state  # reps: (mbsz, hdim)
+
+    def calculate_loss_and_cache_gradients(self, reps, masks) -> Tensor:
+        """Calculate the cross-entropy loss and cache the gradients wrt. the embeddings."""
+        # we want partial grads on all the chunked embeddings
+        loss = self.calculate_loss(reps, masks, with_backward=True)
+        loss = loss.detach().requires_grad_()
+
+        self.cache = [
+            [r.grad for r in rs] for rs in reps
+        ]  # e.g. 3 * bsz/mbsz * (mbsz, hdim)
+
+        return loss
+
+    def calculate_loss(self, reps, masks, with_backward: bool = False) -> Tensor:
+        """Calculate the cross-entropy loss. No need to cache the gradients."""
+        # Each sub-list in reps is a list of mini-batch chunk embeddings
```

Comment:
You could add those comments to the docstring or to the parameters definitions so it's clear from our documentation, the user would not have to dig the code to find those informations